From 9e5be0899f60208046f1c5e2d32a2dd207d8fa00 Mon Sep 17 00:00:00 2001
From: Tanu Kaskinen <tanuk@iki.fi>
Date: Thu, 24 May 2018 20:31:51 +0300
Subject: [PATCH] alsa-card: fix null dereference

jack->melem can be null if the jack disappears between probing the card
and the init_jacks() call. I don't know if jacks actually ever disappear
like that (seems unlikely), but this check is in any case needed as long
as init_jacks() has proper handling for the jack disappearing case
(rather than just an assert).

There was a crash report[1] that indicated that card_suspend_changed()
called report_jack_state() with a null melem. I don't know if that was
because the jack actually disappeared, or is there some other bug too.

[1] https://bugs.freedesktop.org/show_bug.cgi?id=104385

Bug: https://bugs.freedesktop.org/show_bug.cgi?id=104385
Bug-Ubuntu: https://bugs.launchpad.net/bugs/1663528
Forwarded: not-needed
Last-Update: 2019-07-01

Index: pulseaudio/src/modules/alsa/module-alsa-card.c
===================================================================
--- pulseaudio.orig/src/modules/alsa/module-alsa-card.c
+++ pulseaudio/src/modules/alsa/module-alsa-card.c
@@ -625,7 +625,8 @@ static pa_hook_result_t card_suspend_cha
     if (card->suspend_cause == 0) {
         /* We were unsuspended, update jack state in case it changed while we were suspended */
         PA_HASHMAP_FOREACH(jack, u->jacks, state) {
-            report_jack_state(jack->melem, 0);
+            if (jack->melem)
+                report_jack_state(jack->melem, 0);
         }
     }
 
